### PR TITLE
🌱 Test rolling update scale-in doesn't race machine deletion

### PR DESCRIPTION
**Pull request by macaptain**
**What this PR does / why we need it**: 
I was curious whether whether the following scenario was possible: 
During a rolling update, after the MachineDeployment reconciler has decremented the old MachineSet Spec.Replicas, but before the Machine has been fully destroyed, the next loop of reconciliation scales up the new MachineSet. 

If this were possible, a rolling update could temporarily consume more resources in the MachineDeployment than Replicas + MaxSurge.

A comment on `TotalMachineSetsReplicaSum` suggests this is guaranteed not to happen: 

https://github.com/kubernetes-sigs/cluster-api/blob/1e13f448ed5823e26bc98bd48d77f1037e0d443b/internal/controllers/machinedeployment/mdutil/util.go#L456

Therefore, I've added a test covering the case where the old MachineSet has been scaled down, but Status.Replicas has not yet caught up (it would be waiting on MachineSet's reconciler's `waitForMachineDeletion`), and verified that the MachineDeployment reconciler does not attempt to scale up the new MachineSet. This scenario wasn't covered before, but the test passes, so we're all good.

**Which issue(s) this PR fixes**: Let me know if you want an issue raised for this small change.

### CODE REVIEW

File: internal/controllers/machinedeployment/machinedeployment_rolling_test.go [hunk 1]
```diff
@@ -173,6 +173,50 @@ func TestReconcileNewMachineSet(t *testing.T) {
 			},
 			error: nil,
 		},
+		{
+			name: "RollingUpdate strategy: while machine in old machine set is being deleted, machine in new machineset is not scaled in",
```
Comment (enxebre) [inline internal/controllers/machinedeployment/machinedeployment_rolling_test.go:177]: did you mean "machine in new machineset is not scaled **out**"?

To be consistent with other test casas, how about: "RollingUpdate strategy: Scale up accounts for deleting Machines to honour maxSurge"?

**Comment by enxebre** [comment]
Thanks! Nice addition, just a minor comment on the test case description, otherwise lgtm.

File: internal/controllers/machinedeployment/machinedeployment_rolling_test.go (hunk 1 above)
Comment (macaptain) [reply internal/controllers/machinedeployment/machinedeployment_rolling_test.go]: I think 'scale out' is correct when talking about an individual machine, so I did mean 'scale out' here. But the terminology in all comments refers to 'scale up / scale down' a MachineSet, so let's match that. 

Your test case name is more succinct and clearer, thanks! I've pushed that suggestion.

**Comment by enxebre** [comment]
thanks!
/lgtm
/approve

**Pull request merged**